src/transformers/sagemaker/trainer_sm.py

```diff
@@ -126,12 +132,123 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
+            torch.save(self.args, os.path.join(output_dir, "training_args.bin"))
+
+    def _save_checkpoint(self, model, trial, metrics=None):
+        if self.is_model_parallel_enabled:
+            if smp.dp_rank() != 0:
+                return
+
+            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+
+            run_dir = self.args.output_dir
+            self.store_flos()
+
+            output_dir = os.path.join(run_dir, checkpoint_folder)
+            self.save_model(output_dir)
+            # Consolidate the state dict on all processed of dp_rank 0
+            opt_state_dict = self.optimizer.state_dict()
+            # Save it and the scheduler on the main process
+            if self.is_world_process_zero():
+                torch.save(opt_state_dict, os.path.join(output_dir, "optimizer.pt"))
+                with warnings.catch_warnings(record=True) as caught_warnings:
+                    torch.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
+                reissue_pt_warnings(caught_warnings)
+
+            # Determine the new best metric / best model checkpoint
+            if metrics is not None and self.args.metric_for_best_model is not None:
+                metric_to_check = self.args.metric_for_best_model
+                if not metric_to_check.startswith("eval_"):
+                    metric_to_check = f"eval_{metric_to_check}"
+                metric_value = metrics[metric_to_check]
+
+                operator = np.greater if self.args.greater_is_better else np.less
+                if (
+                    self.state.best_metric is None
+                    or self.state.best_model_checkpoint is None
+                    or operator(metric_value, self.state.best_metric)
+                ):
+                    self.state.best_metric = metric_value
+                    self.state.best_model_checkpoint = output_dir
+
+            # Save the Trainer state
+            if self.is_world_process_zero():
+                self.state.save_to_json(os.path.join(output_dir, "trainer_state.json"))
+
+            # Maybe delete some older checkpoints.
+            if self.is_world_process_zero():
+                self._rotate_checkpoints(use_mtime=True, output_dir=run_dir)
+        else:
+            super()._save_checkpoint(self, model, trial, metrics=metrics)
+
+    def _load_optimizer_and_scheduler(self, checkpoint):
+        """If optimizer and scheduler states exist, load them."""
+        if self.is_model_parallel_enabled:
+            if checkpoint is None:
+                return
+
+            if os.path.isfile(os.path.join(checkpoint, "optimizer.pt")) and os.path.isfile(
+                os.path.join(checkpoint, "scheduler.pt")
+            ):
+                self.optimizer.load_state_dict(
+                    torch.load(os.path.join(checkpoint, "optimizer.pt"), map_location="cpu")
```

Comment:
The method is overloaded for this particular line (needs to be loaded on the CPU and not the device).